File: crates/finance/src/mail/zip.rs
//! A zip written here: stored entries, no compression, UTF-8 names. The
//! browser writes the accountant's download the same way (`ui/finances/src/
//! lib/zip.ts`), so the bundle a mail carries is byte-for-byte the shape the
//! download has. PDFs do not compress; a dependency would buy nothing.

/// One file inside the zip.
#[derive(Debug, Clone)]
pub struct Entry {
    /// The path inside the zip, `/`-separated.
    pub name: String,
    /// The bytes.
    pub bytes: Vec<u8>,
}

/// Local-header and central-directory flag: names are UTF-8.
const UTF8: u16 = 0x0800;
/// Version 2.0: stored entries, nothing newer.
const VERSION: u16 = 20;

/// The zip's bytes, entries in the order given.
///
/// # Errors
/// An empty or duplicate name; a name with a backslash or a leading slash;
/// more than 65,535 entries or a size the format cannot carry.
pub fn write(entries: &[Entry]) -> Result<Vec<u8>, String> {
    if entries.len() > usize::from(u16::MAX) {
        return Err("too many files for one zip".into());
    }
    let mut seen = std::collections::HashSet::new();
    for e in entries {
        if e.name.is_empty() || e.name.starts_with('/') || e.name.contains('\\') {
            return Err(format!("not a zip entry name: {:?}", e.name));
        }
        if !seen.insert(e.name.as_str()) {
            return Err(format!("the zip would hold {:?} twice", e.name));
        }
    }
    let (time, date) = dos_now();
    let mut out = Vec::new();
    let mut central = Vec::new();
    for e in entries {
        let name = e.name.as_bytes();
        let size = u32::try_from(e.bytes.len())
            .map_err(|_| format!("{} is too large for a zip", e.name))?;
        let name_len =
            u16::try_from(name.len()).map_err(|_| format!("{} is too long a name", e.name))?;
        let crc = crc32fast::hash(&e.bytes);
        let offset = u32::try_from(out.len()).map_err(|_| "the zip is too large".to_owned())?;

        out.extend_from_slice(&0x0403_4b50_u32.to_le_bytes());
        out.extend_from_slice(&VERSION.to_le_bytes());
        out.extend_from_slice(&UTF8.to_le_bytes());
        out.extend_from_slice(&0_u16.to_le_bytes()); // stored
        out.extend_from_slice(&time.to_le_bytes());
        out.extend_from_slice(&date.to_le_bytes());
        out.extend_from_slice(&crc.to_le_bytes());
        out.extend_from_slice(&size.to_le_bytes());
        out.extend_from_slice(&size.to_le_bytes());
        out.extend_from_slice(&name_len.to_le_bytes());
        out.extend_from_slice(&0_u16.to_le_bytes()); // extra
        out.extend_from_slice(name);
        out.extend_from_slice(&e.bytes);

        central.extend_from_slice(&0x0201_4b50_u32.to_le_bytes());
        central.extend_from_slice(&VERSION.to_le_bytes()); // made by
        central.extend_from_slice(&VERSION.to_le_bytes()); // needed
        central.extend_from_slice(&UTF8.to_le_bytes());
        central.extend_from_slice(&0_u16.to_le_bytes()); // stored
        central.extend_from_slice(&time.to_le_bytes());
        central.extend_from_slice(&date.to_le_bytes());
        central.extend_from_slice(&crc.to_le_bytes());
        central.extend_from_slice(&size.to_le_bytes());
        central.extend_from_slice(&size.to_le_bytes());
        central.extend_from_slice(&name_len.to_le_bytes());
        central.extend_from_slice(&0_u16.to_le_bytes()); // extra
        central.extend_from_slice(&0_u16.to_le_bytes()); // comment
        central.extend_from_slice(&0_u16.to_le_bytes()); // disk
        central.extend_from_slice(&0_u16.to_le_bytes()); // internal attributes
        central.extend_from_slice(&0_u32.to_le_bytes()); // external attributes
        central.extend_from_slice(&offset.to_le_bytes());
        central.extend_from_slice(name);
    }
    let central_offset = u32::try_from(out.len()).map_err(|_| "the zip is too large".to_owned())?;
    let central_size =
        u32::try_from(central.len()).map_err(|_| "the zip is too large".to_owned())?;
    #[allow(clippy::cast_possible_truncation)] // checked above
    let count = entries.len() as u16;
    out.extend_from_slice(&central);
    out.extend_from_slice(&0x0605_4b50_u32.to_le_bytes());
    out.extend_from_slice(&0_u16.to_le_bytes()); // this disk
    out.extend_from_slice(&0_u16.to_le_bytes()); // directory's disk
    out.extend_from_slice(&count.to_le_bytes());
    out.extend_from_slice(&count.to_le_bytes());
    out.extend_from_slice(&central_size.to_le_bytes());
    out.extend_from_slice(&central_offset.to_le_bytes());
    out.extend_from_slice(&0_u16.to_le_bytes()); // comment
    Ok(out)
}

/// MS-DOS time and date fields for now, as the format wants them.
fn dos_now() -> (u16, u16) {
    use chrono::{Datelike, Timelike};
    let now = chrono::Local::now();
    let year = u16::try_from((now.year() - 1980).max(0))
        .unwrap_or(0)
        .min(127);
    #[allow(clippy::cast_possible_truncation)] // all fields are small by construction
    let (month, day, hour, minute, second) = (
        now.month() as u16,
        now.day() as u16,
        now.hour() as u16,
        now.minute() as u16,
        now.second() as u16,
    );
    (
        (hour << 11) | (minute << 5) | (second >> 1),
        (year << 9) | (month << 5) | day,
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Reads the zip back with nothing but the format: every local header,
    /// then the directory, and the counts agree.
    fn entries_of(zip: &[u8]) -> Vec<(String, Vec<u8>, u32)> {
        let end = zip.len() - 22;
        assert_eq!(&zip[end..end + 4], &0x0605_4b50_u32.to_le_bytes());
        let count = u16::from_le_bytes([zip[end + 8], zip[end + 9]]);
        let dir_size = u32::from_le_bytes(zip[end + 12..end + 16].try_into().unwrap()) as usize;
        let dir_at = u32::from_le_bytes(zip[end + 16..end + 20].try_into().unwrap()) as usize;
        assert_eq!(dir_at + dir_size, end);
        let mut out = Vec::new();
        let mut at = dir_at;
        for _ in 0..count {
            assert_eq!(&zip[at..at + 4], &0x0201_4b50_u32.to_le_bytes());
            let crc = u32::from_le_bytes(zip[at + 16..at + 20].try_into().unwrap());
            let size = u32::from_le_bytes(zip[at + 20..at + 24].try_into().unwrap()) as usize;
            let name_len = u16::from_le_bytes([zip[at + 28], zip[at + 29]]) as usize;
            let offset = u32::from_le_bytes(zip[at + 42..at + 46].try_into().unwrap()) as usize;
            let name = String::from_utf8(zip[at + 46..at + 46 + name_len].to_vec()).unwrap();
            assert_eq!(&zip[offset..offset + 4], &0x0403_4b50_u32.to_le_bytes());
            let data_at = offset + 30 + name_len;
            out.push((name, zip[data_at..data_at + size].to_vec(), crc));
            at += 46 + name_len;
        }
        out
    }

    #[test]
    fn a_zip_holds_its_entries_in_order_with_their_crcs() {
        let zip = write(&[
            Entry {
                name: "PROCITAJ.txt".into(),
                bytes: b"hello".to_vec(),
            },
            Entry {
                name: "racuni/2026-08-06_Hetzner_55_00_EUR.pdf".into(),
                bytes: b"%PDF-1.4".to_vec(),
            },
            Entry {
                name: "racuni/prazno.pdf".into(),
                bytes: vec![],
            },
        ])
        .unwrap();
        let got = entries_of(&zip);
        assert_eq!(got.len(), 3);
        assert_eq!(got[0].0, "PROCITAJ.txt");
        assert_eq!(got[0].1, b"hello");
        assert_eq!(got[0].2, 0x3610_a686, "the CRC-32 of \"hello\"");
        assert_eq!(got[1].0, "racuni/2026-08-06_Hetzner_55_00_EUR.pdf");
        assert_eq!(got[1].1, b"%PDF-1.4");
        assert_eq!(got[2].1, b"");
        assert_eq!(got[2].2, 0);
        assert!(
            write(&[]).unwrap().len() == 22,
            "an empty zip is the end record alone"
        );
    }

    #[test]
    fn a_bad_name_is_refused_before_anything_is_written() {
        let e = |name: &str| Entry {
            name: name.into(),
            bytes: vec![],
        };
        assert!(
            write(&[e("")])
                .unwrap_err()
                .contains("not a zip entry name")
        );
        assert!(
            write(&[e("/abs")])
                .unwrap_err()
                .contains("not a zip entry name")
        );
        assert!(
            write(&[e("a\\b")])
                .unwrap_err()
                .contains("not a zip entry name")
        );
        assert!(write(&[e("x"), e("x")]).unwrap_err().contains("twice"));
    }
}

File: ui/finances/src/lib/zip.ts
// A zip file, written in the browser: stored entries, no compression, UTF-8
// names. Enough for a bundle of PDFs the accountant will open, and small
// enough not to warrant a dependency.

const CRC_TABLE = (() => {
  const t = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    t[n] = c >>> 0;
  }
  return t;
})();

function crc32(bytes: Uint8Array): number {
  let c = 0xffffffff;
  for (let i = 0; i < bytes.length; i++) c = CRC_TABLE[(c ^ bytes[i]!) & 0xff]! ^ (c >>> 8);
  return (c ^ 0xffffffff) >>> 0;
}

/** MS-DOS date and time fields, as the zip format wants them. */
function dosDateTime(d: Date): { date: number; time: number } {
  const date = ((d.getFullYear() - 1980) << 9) | ((d.getMonth() + 1) << 5) | d.getDate();
  const time = (d.getHours() << 11) | (d.getMinutes() << 5) | (d.getSeconds() >> 1);
  return { date, time };
}

export type ZipEntry = { name: string; bytes: Uint8Array; modified?: Date };

/** The zip's bytes, entries in the order given. Names must be unique. */
export function zip(entries: ZipEntry[]): Blob {
  const encoder = new TextEncoder();
  const parts: Uint8Array[] = [];
  const central: Uint8Array[] = [];
  let offset = 0;
  for (const e of entries) {
    const name = encoder.encode(e.name);
    const crc = crc32(e.bytes);
    const { date, time } = dosDateTime(e.modified ?? new Date());
    const local = new DataView(new ArrayBuffer(30));
    local.setUint32(0, 0x04034b50, true);
    local.setUint16(4, 20, true); // version needed
    local.setUint16(6, 0x0800, true); // UTF-8 names
    local.setUint16(8, 0, true); // stored
    local.setUint16(10, time, true);
    local.setUint16(12, date, true);
    local.setUint32(14, crc, true);
    local.setUint32(18, e.bytes.length, true);
    local.setUint32(22, e.bytes.length, true);
    local.setUint16(26, name.length, true);
    local.setUint16(28, 0, true);
    parts.push(new Uint8Array(local.buffer), name, e.bytes);

    const c = new DataView(new ArrayBuffer(46));
    c.setUint32(0, 0x02014b50, true);
    c.setUint16(4, 20, true); // made by
    c.setUint16(6, 20, true); // needed
    c.setUint16(8, 0x0800, true);
    c.setUint16(10, 0, true);
    c.setUint16(12, time, true);
    c.setUint16(14, date, true);
    c.setUint32(16, crc, true);
    c.setUint32(20, e.bytes.length, true);
    c.setUint32(24, e.bytes.length, true);
    c.setUint16(28, name.length, true);
    c.setUint16(30, 0, true); // extra
    c.setUint16(32, 0, true); // comment
    c.setUint16(34, 0, true); // disk
    c.setUint16(36, 0, true); // internal attrs
    c.setUint32(38, 0, true); // external attrs
    c.setUint32(42, offset, true);
    central.push(new Uint8Array(c.buffer), name);
    offset += 30 + name.length + e.bytes.length;
  }
  const centralSize = central.reduce((n, p) => n + p.length, 0);
  const end = new DataView(new ArrayBuffer(22));
  end.setUint32(0, 0x06054b50, true);
  end.setUint16(4, 0, true);
  end.setUint16(6, 0, true);
  end.setUint16(8, entries.length, true);
  end.setUint16(10, entries.length, true);
  end.setUint32(12, centralSize, true);
  end.setUint32(16, offset, true);
  end.setUint16(20, 0, true);
  // One contiguous buffer: Blob wants ArrayBuffer-backed views, and a
  // Uint8Array from a base64 decode may be typed as ArrayBufferLike.
  const all = [...parts, ...central, new Uint8Array(end.buffer)];
  const out = new Uint8Array(new ArrayBuffer(all.reduce((n, p) => n + p.length, 0)));
  let at = 0;
  for (const p of all) {
    out.set(p, at);
    at += p.length;
  }
  return new Blob([out], { type: "application/zip" });
}

/** A file name that is safe inside the zip and readable outside it. */
export function safeName(s: string): string {
  return (
    s
      .normalize("NFKD")
      .replace(/[̀-ͯ]/g, "")
      .replace(/[^A-Za-z0-9._-]+/g, "_")
      .replace(/^_+|_+$/g, "")
      .slice(0, 60) || "file"
  );
}
